### ISSUE TITLE
Incorrect handling of bi-directional relationship In Neo4JClient

### ISSUE THREAD

**Issue by ghost** [Impetus/kundera#581]
Current code:

```
         // entity into Map field for target entity
                        Field bidirectionalField = relation.getBiDirectionalField();
                        Map<Object, Object> sourceEntitiesMap = new HashMap<Object, Object>();
                        if (bidirectionalField != null)
                        {
                            for (Relationship incomingRelationship : endNode.getRelationships(Direction.INCOMING,
                                    DynamicRelationshipType.withName(relation.getJoinColumnName(kunderaMetadata))))
                            {
                                Node startNode = incomingRelationship.getStartNode();
                                Object sourceEntity = nodeIdToEntityMap.get(startNode.getId());
                                if (sourceEntity == null)
                                {
                                    sourceEntity = mapper.getEntityFromNode(startNode, m);
                                    nodeIdToEntityMap.put(startNode.getId(), sourceEntity);
                                }
                                sourceEntitiesMap.put(relationshipEntity, sourceEntity);
                            }
                            PropertyAccessorHelper.set(targetEntity, bidirectionalField, sourceEntitiesMap);
                        }

```

Currently the way neo4j handles bi-directional relationship is based just on type, which in my opinion is incorrect. Given following scenario:

```
Person 
  name: "Tim Cook"
  memberOf -> Organization (APPLE)

Organiztion:
   name: "Apple"
   founder -> Person (Steve Jobs)

Person:
   name: "Steve Jobs"
   ....

```

Now let's say you load the the person with name "Time Cook" then the code above would actually creates a relationship like this

Person("Tim Cook") -- memberOf -> Organization ("Applie") -- founder -->  Person("Tim Cook")

which is incorrect, the point here is that bi-directionality can't really be decided based on just type here, so, 
1. May be there's no need to create a bi-directional relationship like this? since we already have the relationship entity that takes care of this?
   Or 
2. The Code can be fixed so that name of the relationship is checked along with the type, somewhat like the following code. 

```
                        MapKeyJoinColumn incomingRel = bidirectionalField.getAnnotation(MapKeyJoinColumn.class);
                        DynamicRelationshipType outgoingRel =  DynamicRelationshipType.withName(relation.getJoinColumnName(kunderaMetadata));
                        if (bidirectionalField != null && incomingRel.name().equals(outgoingRel.name()))
                        {
                             ....
                         }
```

The code above not only checks for the field type but also checks to see if the relationship name matches by extracting that information from  MapKeyJoinColumn information.

**Comment by ghost** [comment]
Ok, so i guess the fix might actually need to be done while building relation metadata object, so inside relation class when setting the reference class

```
    public void setBiDirectionalField(Class referencedClass) {
        Field[] fields = this.getTargetEntity().getDeclaredFields();
        Class<?> clazzz = null;
        for (Field field : fields) {
            if (!ReflectUtils.isTransientOrStatic(field)) {
                clazzz = field.getType();
                String mappedBy = mappedBy(clazzz);
                if (PropertyAccessorHelper.isCollection(clazzz)) {
                    ParameterizedType type = (ParameterizedType) field.getGenericType();
                    Type[] types = type.getActualTypeArguments();
                    clazzz = (Class<?>) types[0];
                } else if (Map.class.isAssignableFrom(clazzz)) {
                    ParameterizedType type = (ParameterizedType) field.getGenericType();
                    Type[] types = type.getActualTypeArguments();
                    clazzz = (Class<?>) types[1];
                }
           //HERE ALONG WITH CHECKING CLASS TYPE CHECK THE MAPPED BY ATTRIBUTE (IF ONE EXISTS).
                if ((mappedBy == null || mappedBy.equalsIgnoreCase(field.getName())) && clazzz.equals(referencedClass)) {
                    biDirectionalField = field;
                    break;
                }
            }
        }

    }

 // NEW METHOD
    private String mappedBy(Class<?> clazzz) {
        ManyToMany m2m = clazzz.getAnnotation(ManyToMany.class);
        return m2m != null ? m2m.mappedBy() : null;
    }
}

```

**Comment by kkmishra** [comment]
Hi @mbellani360 

Are you trying self association in Neo4j?

Thanks
Kuldeep

**Comment by ghost** [comment]
I'm modelling  http://schema.org/Person, this is not a self association.

**Comment by kkmishra** [comment]
In given modeling definition http://schema.org/Person.
It seems to me that person has some self-association.
Like person has children, colleague, follows etc.

Correct me if I am getting something wrong!

**Comment by kkmishra** [comment]
Hi @mbellani360 

In order to quickly resolve the issue it will be good to connect once, can you please share your gmail id or any other messenger id for further discussion? Or you can also connect me at kuld.cs.mishra@gmail.com.

Thanks
Kuldeep

**Comment by kkmishra** [comment]
A test case has been added into current trunk branch.
Please follow the following link and verify your scenario.

https://github.com/impetus-opensource/Kundera/blob/trunk/src/kundera-neo4j/src/test/java/com/impetus/client/neo4j/imdb/ActorWithMultipleRelationTest.java

**Comment by chhavigangwal** [comment]
Fixed releasing with 2.13.
